File: ts/indicators/john-ehlers/sine-wave/sine-wave.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, DefaultBarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { Band } from '../../core/outputs/band';
import { DominantCycle } from '../dominant-cycle/dominant-cycle';
import { HilbertTransformerCycleEstimatorType } from '../hilbert-transformer/cycle-estimator-type';
import { createEstimator, estimatorMoniker } from '../hilbert-transformer/common';
import { SineWaveParams } from './params';

const DEFAULT_ALPHA_EMA_PERIOD_ADDITIONAL = 0.33;
const DEFAULT_SMOOTHING_LENGTH = 4;
const DEFAULT_ALPHA_EMA_QI = 0.2;
const DEFAULT_ALPHA_EMA_PERIOD = 0.2;
// MBST's DominantCyclePeriod default warm-up is MaxPeriod * 2 = 100.
const DEFAULT_WARM_UP_PERIOD = 100;

const DEG2RAD = Math.PI / 180.0;
const LEAD_OFFSET = 45.0;

/** __Sine Wave__ (Ehlers) computes a clear sine wave representation of the dominant cycle phase.
 *
 * It exposes five outputs:
 *
 *	- Value: the sine wave value, sin(phase·Deg2Rad).
 *	- Lead: the sine wave lead value, sin((phase+45)·Deg2Rad).
 *	- Band: a band with Value as the upper line and Lead as the lower line.
 *	- DominantCyclePeriod: the smoothed dominant cycle period.
 *	- DominantCyclePhase: the dominant cycle phase, in degrees.
 *
 * Reference:
 *
 *	John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 95-105.
 */
export class SineWave implements Indicator {
  private readonly dc: DominantCycle;
  private primed = false;
  private value = Number.NaN;
  private lead = Number.NaN;

  private readonly mnemonicValue: string;
  private readonly descriptionValue: string;
  private readonly mnemonicLead: string;
  private readonly descriptionLead: string;
  private readonly mnemonicBand: string;
  private readonly descriptionBand: string;
  private readonly mnemonicDCP: string;
  private readonly descriptionDCP: string;
  private readonly mnemonicDCPhase: string;
  private readonly descriptionDCPhase: string;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  /** Creates an instance using default parameters (α=0.33, HomodyneDiscriminator cycle
   * estimator with smoothingLength=4, αq=0.2, αp=0.2, warmUpPeriod=100, BarComponent.Median). */
  public static default(): SineWave {
    return new SineWave({
      alphaEmaPeriodAdditional: DEFAULT_ALPHA_EMA_PERIOD_ADDITIONAL,
      estimatorType: HilbertTransformerCycleEstimatorType.HomodyneDiscriminator,
      estimatorParams: {
        smoothingLength: DEFAULT_SMOOTHING_LENGTH,
        alphaEmaQuadratureInPhase: DEFAULT_ALPHA_EMA_QI,
        alphaEmaPeriod: DEFAULT_ALPHA_EMA_PERIOD,
        warmUpPeriod: DEFAULT_WARM_UP_PERIOD,
      },
    });
  }

  /** Creates an instance based on the given parameters. */
  public static fromParams(params: SineWaveParams): SineWave {
    return new SineWave(params);
  }

  private constructor(params: SineWaveParams) {
    const alpha = params.alphaEmaPeriodAdditional;
    if (alpha <= 0 || alpha > 1) {
      throw new Error('invalid sine wave parameters: α for additional smoothing should be in range (0, 1]');
    }

    // SineWave defaults to BarComponent.Median (not the framework default), so it always
    // shows in the mnemonic.
    const bc = params.barComponent ?? BarComponent.Median;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    // Inner DominantCycle built with explicit components so its mnemonic contains hl/2.
    this.dc = DominantCycle.fromParams({
      alphaEmaPeriodAdditional: alpha,
      estimatorType: params.estimatorType,
      estimatorParams: params.estimatorParams,
      barComponent: bc,
      quoteComponent: qc,
      tradeComponent: tc,
    });

    // Compose estimator moniker (same logic as DominantCycle).
    const effectiveType = params.estimatorType ?? HilbertTransformerCycleEstimatorType.HomodyneDiscriminator;
    const htce = createEstimator(params.estimatorType, params.estimatorParams);
    let em = '';
    const isDefaultHd = effectiveType === HilbertTransformerCycleEstimatorType.HomodyneDiscriminator
      && htce.smoothingLength === DEFAULT_SMOOTHING_LENGTH
      && htce.alphaEmaQuadratureInPhase === DEFAULT_ALPHA_EMA_QI
      && htce.alphaEmaPeriod === DEFAULT_ALPHA_EMA_PERIOD;
    if (!isDefaultHd) {
      const moniker = estimatorMoniker(effectiveType, htce);
      if (moniker.length > 0) {
        em = ', ' + moniker;
      }
    }

    const cm = componentTripleMnemonic(bc, qc, tc);
    const a = alpha.toFixed(3);

    this.mnemonicValue = `sw(${a}${em}${cm})`;
    this.mnemonicLead = `sw-lead(${a}${em}${cm})`;
    this.mnemonicBand = `sw-band(${a}${em}${cm})`;
    this.mnemonicDCP = `dcp(${a}${em}${cm})`;
    this.mnemonicDCPhase = `dcph(${a}${em}${cm})`;

    this.descriptionValue = 'Sine wave ' + this.mnemonicValue;
    this.descriptionLead = 'Sine wave lead ' + this.mnemonicLead;
    this.descriptionBand = 'Sine wave band ' + this.mnemonicBand;
    this.descriptionDCP = 'Dominant cycle period ' + this.mnemonicDCP;
    this.descriptionDCPhase = 'Dominant cycle phase ' + this.mnemonicDCPhase;

    // Silence unused-variable warning for DefaultBarComponent import (kept for consistency).
    void DefaultBarComponent;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean { return this.primed; }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.SineWave,
      this.mnemonicValue,
      this.descriptionValue,
      [
        { mnemonic: this.mnemonicValue, description: this.descriptionValue },
        { mnemonic: this.mnemonicLead, description: this.descriptionLead },
        { mnemonic: this.mnemonicBand, description: this.descriptionBand },
        { mnemonic: this.mnemonicDCP, description: this.descriptionDCP },
        { mnemonic: this.mnemonicDCPhase, description: this.descriptionDCPhase },
      ],
    );
  }

  /** Updates the indicator given the next sample value. Returns the quadruple
   * (value, lead, period, phase). Returns all-NaN if not yet primed. */
  public update(sample: number): [number, number, number, number] {
    if (Number.isNaN(sample)) {
      return [sample, sample, sample, sample];
    }

    const [, period, phase] = this.dc.update(sample);

    if (Number.isNaN(phase)) {
      return [Number.NaN, Number.NaN, Number.NaN, Number.NaN];
    }

    this.primed = true;
    this.value = Math.sin(phase * DEG2RAD);
    this.lead = Math.sin((phase + LEAD_OFFSET) * DEG2RAD);
    return [this.value, this.lead, period, phase];
  }

  /** Updates an indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    return this.updateEntity(sample.time, sample.value);
  }

  /** Updates an indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    return this.updateEntity(sample.time, this.barComponentFunc(sample));
  }

  /** Updates an indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    return this.updateEntity(sample.time, this.quoteComponentFunc(sample));
  }

  /** Updates an indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    return this.updateEntity(sample.time, this.tradeComponentFunc(sample));
  }

  private updateEntity(time: Date, sample: number): IndicatorOutput {
    const [value, lead, period, phase] = this.update(sample);

    const sv = new Scalar();
    sv.time = time;
    sv.value = value;

    const sl = new Scalar();
    sl.time = time;
    sl.value = lead;

    const band = new Band();
    band.time = time;
    band.upper = value;
    band.lower = lead;

    const sp = new Scalar();
    sp.time = time;
    sp.value = period;

    const sph = new Scalar();
    sph.time = time;
    sph.value = phase;

    return [sv, sl, band, sp, sph];
  }
}

File: go/indicators/johnehlers/sinewave/sinewave.go
package sinewave

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
	"zpano/indicators/johnehlers/dominantcycle"
	"zpano/indicators/johnehlers/hilberttransformer"
)

// Deg2Rad converts degrees to radians.
const deg2Rad = math.Pi / 180.0

// SineWave is the Ehlers' Sine Wave indicator.
//
// It exposes five outputs:
//
//   - Value: the sine wave value, sin(phase·Deg2Rad).
//   - Lead: the sine wave lead value, sin((phase+45)·Deg2Rad).
//   - Band: a band with Upper=Value and Lower=Lead.
//   - DominantCyclePeriod: the smoothed dominant cycle period.
//   - DominantCyclePhase: the dominant cycle phase, in degrees.
//
// Reference:
// John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 95-105.
type SineWave struct {
	mu                 sync.RWMutex
	mnemonic           string
	description        string
	mnemonicLead       string
	descriptionLead    string
	mnemonicBand       string
	descriptionBand    string
	mnemonicDCP        string
	descriptionDCP     string
	mnemonicDCPhase    string
	descriptionDCPhase string
	dc                 *dominantcycle.DominantCycle
	primed             bool
	value              float64
	lead               float64
	barFunc            entities.BarFunc
	quoteFunc          entities.QuoteFunc
	tradeFunc          entities.TradeFunc
}

// NewSineWaveDefault returns an instance of the indicator
// created using default values of the parameters.
func NewSineWaveDefault() (*SineWave, error) {
	const (
		smoothingLength           = 4
		alphaEmaQuadratureInPhase = 0.2
		alphaEmaPeriod            = 0.2
		alphaEmaPeriodAdditional  = 0.33
		warmUpPeriod              = 100
	)

	return newSineWave(
		hilberttransformer.HomodyneDiscriminator,
		&hilberttransformer.CycleEstimatorParams{
			SmoothingLength:           smoothingLength,
			AlphaEmaQuadratureInPhase: alphaEmaQuadratureInPhase,
			AlphaEmaPeriod:            alphaEmaPeriod,
			WarmUpPeriod:              warmUpPeriod,
		},
		alphaEmaPeriodAdditional,
		0, 0, 0)
}

// NewSineWaveParams returns an instance of the indicator created using supplied parameters.
func NewSineWaveParams(p *Params) (*SineWave, error) {
	return newSineWave(
		p.EstimatorType, &p.EstimatorParams,
		p.AlphaEmaPeriodAdditional,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen
func newSineWave(
	estimatorType hilberttransformer.CycleEstimatorType,
	estimatorParams *hilberttransformer.CycleEstimatorParams,
	alphaEmaPeriodAdditional float64,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*SineWave, error) {
	const (
		invalid    = "invalid sine wave parameters"
		fmta       = "%s: α for additional smoothing should be in range (0, 1]"
		fmtw       = "%s: %w"
		fmtnValue  = "sw(%.3f%s%s)"
		fmtnLead   = "sw-lead(%.3f%s%s)"
		fmtnBand   = "sw-band(%.3f%s%s)"
		fmtnDCP    = "dcp(%.3f%s%s)"
		fmtnDCPha  = "dcph(%.3f%s%s)"
		four       = 4
		alpha      = 0.2
		descrValue = "Sine wave "
		descrLead  = "Sine wave lead "
		descrBand  = "Sine wave band "
		descrDCP   = "Dominant cycle period "
		descrDCPha = "Dominant cycle phase "
	)

	if alphaEmaPeriodAdditional <= 0. || alphaEmaPeriodAdditional > 1. {
		return nil, fmt.Errorf(fmta, invalid)
	}

	// Resolve defaults for component functions.
	// SineWave defaults to BarMedianPrice (not the framework default), so it always shows in the mnemonic.
	if bc == 0 {
		bc = entities.BarMedianPrice
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	// Build the inner DominantCycle with explicit components.
	dcParams := &dominantcycle.Params{
		EstimatorType:            estimatorType,
		EstimatorParams:          *estimatorParams,
		AlphaEmaPeriodAdditional: alphaEmaPeriodAdditional,
		BarComponent:             bc,
		QuoteComponent:           qc,
		TradeComponent:           tc,
	}

	dc, err := dominantcycle.NewDominantCycleParams(dcParams)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Compose the estimator moniker (same logic as DominantCycle).
	estimator, err := hilberttransformer.NewCycleEstimator(estimatorType, estimatorParams)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	estimatorMoniker := ""
	if estimatorType != hilberttransformer.HomodyneDiscriminator ||
		estimatorParams.SmoothingLength != four ||
		estimatorParams.AlphaEmaQuadratureInPhase != alpha ||
		estimatorParams.AlphaEmaPeriod != alpha {
		estimatorMoniker = hilberttransformer.EstimatorMoniker(estimatorType, estimator)
		if len(estimatorMoniker) > 0 {
			estimatorMoniker = ", " + estimatorMoniker
		}
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)

	mnemonicValue := fmt.Sprintf(fmtnValue, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)
	mnemonicLead := fmt.Sprintf(fmtnLead, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)
	mnemonicBand := fmt.Sprintf(fmtnBand, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)
	mnemonicDCP := fmt.Sprintf(fmtnDCP, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)
	mnemonicDCPha := fmt.Sprintf(fmtnDCPha, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)

	barFunc, err := entities.BarComponentFunc(bc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	quoteFunc, err := entities.QuoteComponentFunc(qc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	tradeFunc, err := entities.TradeComponentFunc(tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	nan := math.NaN()

	return &SineWave{
		mnemonic:           mnemonicValue,
		description:        descrValue + mnemonicValue,
		mnemonicLead:       mnemonicLead,
		descriptionLead:    descrLead + mnemonicLead,
		mnemonicBand:       mnemonicBand,
		descriptionBand:    descrBand + mnemonicBand,
		mnemonicDCP:        mnemonicDCP,
		descriptionDCP:     descrDCP + mnemonicDCP,
		mnemonicDCPhase:    mnemonicDCPha,
		descriptionDCPhase: descrDCPha + mnemonicDCPha,
		dc:                 dc,
		value:              nan,
		lead:               nan,
		barFunc:            barFunc,
		quoteFunc:          quoteFunc,
		tradeFunc:          tradeFunc,
	}, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *SineWave) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
func (s *SineWave) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.SineWave,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: s.description},
			{Mnemonic: s.mnemonicLead, Description: s.descriptionLead},
			{Mnemonic: s.mnemonicBand, Description: s.descriptionBand},
			{Mnemonic: s.mnemonicDCP, Description: s.descriptionDCP},
			{Mnemonic: s.mnemonicDCPhase, Description: s.descriptionDCPhase},
		},
	)
}

// Update updates the value of the indicator given the next sample, returning
// the (value, lead, period, phase) tuple. Returns NaN values if the indicator is not yet primed.
func (s *SineWave) Update(sample float64) (float64, float64, float64, float64) {
	if math.IsNaN(sample) {
		return sample, sample, sample, sample
	}

	// Delegate to the inner DominantCycle. Its own mutex is independent.
	_, period, phase := s.dc.Update(sample)

	s.mu.Lock()
	defer s.mu.Unlock()

	if math.IsNaN(phase) {
		nan := math.NaN()

		return nan, nan, nan, nan
	}

	const leadOffset = 45.0

	s.primed = true
	s.value = math.Sin(phase * deg2Rad)
	s.lead = math.Sin((phase + leadOffset) * deg2Rad)

	return s.value, s.lead, period, phase
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *SineWave) UpdateScalar(sample *entities.Scalar) core.Output {
	return s.updateEntity(sample.Time, sample.Value)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *SineWave) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, s.barFunc(sample))
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *SineWave) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, s.quoteFunc(sample))
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *SineWave) UpdateTrade(sample *entities.Trade) core.Output {
	return s.updateEntity(sample.Time, s.tradeFunc(sample))
}

func (s *SineWave) updateEntity(time time.Time, sample float64) core.Output {
	const length = 5

	value, lead, period, phase := s.Update(sample)

	output := make([]any, length)
	i := 0
	output[i] = entities.Scalar{Time: time, Value: value}
	i++
	output[i] = entities.Scalar{Time: time, Value: lead}
	i++
	output[i] = outputs.Band{Time: time, Upper: value, Lower: lead}
	i++
	output[i] = entities.Scalar{Time: time, Value: period}
	i++
	output[i] = entities.Scalar{Time: time, Value: phase}

	return output
}
